### PR TITLE
 [Sortable] ORM Custom Types 

### LINKED ISSUE #1545: [Sortable] ORM Custom Types
We have a one to many assoctions.
One Person has many Cars.
We want to sort Cars by Person.
Everything works fine.
Now, the Person became a custom doctrine type as id.
The get MaxPosition query is broken, cause it doesn't resolve the type.

### PR DESCRIPTION
**Pull request by Gladhon**
We have a one to many assoctions.
One Person has many Cars.
We want to sort Cars by Person ( by group ).
Everything works fine.
Now, the Person became a custom doctrine type as id.
The get MaxPosition query is broken, cause it doesn't resolve the custom type.

Fix #1545

### CODE REVIEW

**Comment by Padam87** [comment]
"Now, the Person became a custom doctrine type as id."

If I understand it correctly your ID is a custom type?

Is that a good idea? I would image that doctrine has a lot a problems resolving the id. You would have to specify the type in every query builder... you will have a lot of problems with 3rd party libs.

**Comment by Gladhon** [comment]
Yes and No. Doctrine it self should have problems with that. 3rd party libs could have problems with it, like this one, but is there any reason to not support doctrine custom types ?

**Comment by Padam87** [comment]
In general, custom, and composit primary keys are a bad practice, but let's put that aside for now.

Your PR assumes that every object that is passed as a group is an entity, with a single primary key, and that is why the test are failing... Objects can be \DateTime, or even a custom value object for a custom type, so you must handle those cases as well.

File: lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php [hunk 1]
```diff
@@ -39,12 +39,38 @@ private function addGroupWhere(QueryBuilder $qb, $groups)
...
         }
     }
 
+    /**
+     * @param $value
+     * @return mixed
+     */
+    private function getGroupValue($value)
+    {
+        if (!is_object($value)) {
```
Comment (Padam87) [inline lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php]: is_object check is not enough, you will have to create an `isEntity` function to check that the object is a mapped entity.

File: lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php [hunk 2]
```diff
@@ -39,12 +39,38 @@ private function addGroupWhere(QueryBuilder $qb, $groups)
...
+            return $value;
+        }
+        return $this->getObjectManager()->getUnitOfWork()->getSingleIdentifierValue($value);
+    }
+
+    /**
+     * @param $value
+     * @return \Doctrine\DBAL\Types\Type|null|string
+     */
+    private function getGroupType($value)
```
Comment (Padam87) [inline lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php]: `$group` should be passed as well, and the metadata should be fetched like this:
- if the object is a simple value object, `$metaData->getTypeOfField($group)`
- if the object is an entity, your code is ok

**Comment by l3pp4rd** [comment]
Doctrine is automatically parsing the **identifier** if you pass in the entity in query. It also manages scalar type transformations. What you need is to register a DBAL type for your custom value, because extensions will handle it only the standard way (otherwise there would be many corner cases). Custom types - custom transformations which you need to handle yourself in this case, or open an issue in doctrine ORM because it should extract id from an entity (but though you still need a DBAL type)

**Comment by Gladhon** [comment]
@l3pp4rd 
the custom dbal type is of course registered, but this will not work on query builder.

**Comment by l3pp4rd** [comment]
have you analyzed why exactly it does not work? I think first we need to know the reason

**Comment by Gladhon** [comment]
well i can not see where in doctrine the type should be handled autoamtically - can you show me that

**Comment by l3pp4rd** [comment]
look how it maps [query parameters](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Query.php#L321) to dbal types

**Comment by Gladhon** [comment]
I see, it use ParameterTypeInferer::inferType($value) to get undefinied types , which guess type by value, and not support any custom types. Do you think there have to  custom types supported also ?

**Comment by l3pp4rd** [comment]
well, you could ask this question in ORM issue tracker, because we just can't make our own type guessing implementation. it should be compatible with ORM. you could debug more how your custom parameter is processed and see exactly whether you can adapt it or not

**Comment by Gladhon** [comment]
I would ask that.
On the other hand, setParameter is designed to became the type of the parameter. And i think reading type from classMetadata is not incompatible and also no guessing.

**Comment by l3pp4rd** [comment]
if it reads successfully from the metadata, then you should not have this problem, or am I missing something?

**Comment by Gladhon** [comment]
Yes, if it reads successfully from the metadata, then i should not have this problem.

**Comment by Gladhon** [comment]
@Padam87 
I have no a better check if it's an entity.
I'im not using $metaData->getTypeOfField($group) right now, cause it doesn't have anything to do with my problem. And i dont  likte to change the bahavior from bevore accidentally. But if you wish, i can add that additional

**Comment by l3pp4rd** [comment]
@Gladhon please create a test case first, without it, we cannot merge and find the best solution.

**Comment by Gladhon** [comment]
i dont know where and how to write test the Gedmo\Sortable\Mapping\Event\Adapter\ORM class. I don't find any tests for that adapter. 
I could just write a unit test, that mocked doctrine classes will be called, but it don't seems to make sense

**Comment by l3pp4rd** [comment]
create functional test where fixtures has custom type **Id** like [SortableGroupTest](https://github.com/Atlantic18/DoctrineExtensions/blob/master/tests/Gedmo/Sortable/SortableGroupTest.php)

File: lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php [hunk 3]
```diff
@@ -36,15 +37,54 @@ private function addGroupWhere(QueryBuilder $qb, $groups)
...
+    {
+        if (!$this->isEntity($value)) {
+            return null;
```
Comment (Padam87) [inline lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php]: Please return $metaData->getTypeOfField($group) here ;)

Like I said, custom type value objects are also a possibility. Infer type will not be able to resolve those, and the sortable won't work.

Comment (Gladhon) [inline lib/Gedmo/Sortable/Mapping/Event/Adapter/ORM.php]: added

**Comment by l3pp4rd** [comment]
thanks @Gladhon I'm very busy next week, but I'll find some time to verify and check whether there are better ways to fix the issue or not.

**Comment by Gladhon** [comment]
@l3pp4rd  no problem, best would be a doctrine ORM internal solution, or a helper from doctrine ORM. 

On the other hand, merge this is in, would not prevent anybody to implement a better way later, it should be easier, cause you have at least a test for it.

**Comment by l3pp4rd** [comment]
I'm not sure whether it will be backward compatible, that is mainly why I need to have a look

**Comment by Gladhon** [comment]
Mainly $metaData->getTypeOfField would be used instadof ParameterTypeInferer::inferType($value) for group-id's. I would say, this is not a BC, as nobody should count on that internals. But I will wait for your opinion.

**Pull request merged**